### PR TITLE
docs(release): add support Tiers and supported platforms per Tier

### LINKED ISSUE #4448: Decide on supported platforms and test in CI
## Motivation

We should decide which platforms we will officially support, and test it in CI.

### Designs

- [ ] Decide which platforms (OS distribution, CPU, Rust version, C/C++ compiler, packaging system) we want to support
- [ ] Add tests in CI for those that we don't test already

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

As we get near to our Release Candidate, we should decide on which platforms we're going to support and the level of support and guarantees we'd like to ensure for each.

We've deploying with Debian in Docker and build with Ubuntu and MacOS in our CI, but we haven't defined an strategy around these platforms or how we're going to guarantee support, if any.

Fixes #4448

## Solution

- Create a document including Tiers and the guarantees each "must" and "should" ensure for users and internally
- Create a document with the Supported Platforms (based from our actual build and test scenarios) linked to the respective Tier

This documents are based on Rust's approach with Tiers and Supported platforms https://github.com/rust-lang/rust/tree/master/src/doc/rustc/src

## Review

Most people from @ZcashFoundation/zebra-team should have a look at this

### Reviewer Checklist

  - [ ] Verify correct wording is being used
  - [ ] Teams and developer amounts relate to our reality
  - [ ] Platforms and specific libraries relate to our reality
  - [ ] Should we add information on why we dropped `windows` support?

## Follow Up Work

- [ ] Add a resume to README.md targeting `supported-platforms.md` and `target-tier-policies.md`
- [ ] Create a PR including the release process based on these supported platforms

### CODE REVIEW

File: book/src/user/supported-platforms.md
Comment (teor2345) [inline book/src/user/supported-platforms.md]:
```suggestion
quite a good degree and patches are always welcome!
Official builds are not available.
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: There is only one Zebra team, so it's a bit confusing to talk about multiple teams.

Please make these changes wherever a team or teams are mentioned, throughout the document.

```suggestion
determined by the judgment of the Zebra team. Neither this policy nor any
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: We should probably delete this legal wording, please do this throughout the document.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: The current tier 3 targets do not meet this documentation requirement, do we want to delete it?

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: These rules don't make sense for Zebra, we should just say we use the Rust target names.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: Do we want to delete all the requirements for "target maintainers"?

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: This rule doesn't make sense for Zebra.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: All the emulation and cross-compilation requirements don't make sense for Zebra, please delete them in the whole document.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: These stub requirements don't make sense for Zebra, please delete all stub requirements.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: The component part of this requirement doesn't make sense for Zebra, we could say "crates or binaries".

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: The "hello world" part of this requirement doesn't make sense for Zebra.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: This isn't what we do for Zebra - we often block PRs on macOS test failures.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]:
```suggestion
    performance, that alone may
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: This doesn't really make sense for Zebra - we're not likely to use emulation or dedicated hardware.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: It's the "Zcash community", not the Zebra community. Please make this change in the whole document.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]:
```suggestion
- Zebra provides no guarantees about tier 3 targets; they may or may not build.
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]:
```suggestion
on avoiding disruption to ongoing Zebra development.
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: Please make this change throughout the document

```suggestion
Tier 2 and tier 1 targets place work on Zebra developers as a whole, to
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: We might want to re-word this to be more application-focused.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]:
```suggestion
- Tier 2 targets must not leave any significant portions of Zebra unimplemented
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: We don't use RFCs like this for Zebra, can you remove all the references to RFCs in this document?

File: book/src/user/supported-platforms.md
Comment (teor2345) [inline book/src/user/supported-platforms.md]: Where did these kernel and glibc versions come from?
Are they from our GitHub or Docker runners?

(I have similar questions about the other tiers.)

File: book/src/user/supported-platforms.md
Comment (teor2345) [inline book/src/user/supported-platforms.md]: We're missing documentation about what kinds of binaries we're building and distributing for each Tier 1 platform.
Can we add another table that lists the distribution methods?

Currently that's just Docker.

**Review by teor2345 — COMMENTED** [review]
This looks like a good start!

The target tier policy has a lot of rules that apply to the Rust compiler, but not Zebra. I'd suggest deleting those rules first, before making any other changes.

Zebra's project structure and processes are also different, so we'll need to update the remaining rules to match what we actually do. I added a comment the first time I saw each difference.

Since we copied these docs from the Rust project, can we credit them at the end of the doc?

**Comment by str4d** [comment]
We also recently set up platform support policies for `zcashd` based on the Rust ones, which you might find useful to look at for how we adapted them:

- https://zcash.github.io/zcash/user/platform-support.html
- https://zcash.github.io/zcash/dev/platform-tier-policy.html

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: Fixed by [docs(release): remove references to use RFCs](https://github.com/ZcashFoundation/zebra/pull/4773/commits/641a6d08a379dc64ae0b1e8128649056542c1f23)

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: Fixed by: [docs(release): remove text that might generate confusion](https://github.com/ZcashFoundation/zebra/pull/4773/commits/587480aaa7404579a299f3dabf0edc99f59495d8)

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: Fixed by: [docs(release): remove text that might generate confusion](https://github.com/ZcashFoundation/zebra/pull/4773/commits/587480aaa7404579a299f3dabf0edc99f59495d8)

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: Fixed by: [docs(release): remove text that might generate confusion](https://github.com/ZcashFoundation/zebra/pull/4773/commits/587480aaa7404579a299f3dabf0edc99f59495d8)

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: This mentions core engineering and maintainer teams, which don't exist for Zebra.
Can you make these changes throughout the document?

```
  approval of the Zebra team.
```

**Comment by teor2345** [comment]
Thanks for the fixes, let me know when you want me to review this PR again.

File: book/src/user/supported-platforms.md
Comment (gustavovalverde) [reply book/src/user/supported-platforms.md]: Both, but I'll make this more clear with the specific OS.

I think we should use Rust version instead of glibc version

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: Docker in an M1 is being emulated for some platforms. This could apply if we consider this scenario.

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: This seems like a gray area (the text seems valid, but moving to Tier 3 doesn't seems to be an option), let me think it for a while and I'll be back here to discuss it.

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: I'm leaving some parts which refers to "emulation" as we've hit some of the constraints ourself. For example, we wanted to build ARM64 to include support for M1 devices, but the emulation wasn't good enough and took way too much time.

As stated in the Tier 1 platform policy, this applies to us:

>     In particular, if building the platform takes a reasonable amount of time,
>     but the platform cannot run the testsuite in a timely fashion due to low
>     performance of either native code or accurate emulation, that alone may
>     prevent the platform from qualifying as tier 1.


Other statements in this policy start with `If`, so it's not exactly stating a "present" state, it might be a future requirement. The policy is just taking in into consideration so we don't have to change it if new requirements are introduced.

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: We've used emulation before (with ARM64) and might have to use it in the future:
https://github.com/ZcashFoundation/zebra/pull/4773#discussion_r922152609

**Comment by gustavovalverde** [comment]
@ZcashFoundation/general-rust-reviewers this is ready for general review.

@str4d thank you for sharing `zcashd` policies, it helped me improving the actual wording in some pieces.

**Review by conradoplg — COMMENTED** [review]
Looks great!

I don't think we need to add an explanation about Windows, it's documented in tickets if someone wants to dig for information.

I won't approve for now so that others can take a look

File: book/src/user/supported-platforms.md
Comment (teor2345) [inline book/src/user/supported-platforms.md]:
```suggestion
Platforms are identified by their Rust "target triple" which is a string composed by
```

File: book/src/user/supported-platforms.md
Comment (teor2345) [inline book/src/user/supported-platforms.md]: I think it might be easier to read if we make the Rust compiler version into a separate column

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md:16]: "They exist in the codebase" is a Rust compiler-specific note.

These platforms might not even be mentioned in the Zebra codebase.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]: I think we should remove this legal language: "binding agreement or estoppel".

**Review by teor2345 — COMMENTED** [review]
Looks good, thanks for all the updates!

I'll finish reviewing the detailed policy tomorrow.

File: book/src/user/supported-platforms.md
Comment (gustavovalverde) [reply book/src/user/supported-platforms.md]: Actually, a target triple is not a Rust project definition. The target triple is a broader definition which (mainly) comes from the LLVM Compiler Infrastructure Project

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md]: Fixed by [docs: remove legal language](https://github.com/ZcashFoundation/zebra/pull/4773/commits/5b14affdd057db99378a9e104068f711db0fcea8)

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md:16]: Updated here: [docs: rewording tier 3 summary](https://github.com/ZcashFoundation/zebra/pull/4773/commits/9050bf1dcf252611d1a5aed534debc1a4180b5d6)

File: book/src/user/supported-platforms.md
Comment (str4d) [reply book/src/user/supported-platforms.md]: Rust's target triples don't always line up with the LLVM or GCC ones though. We have to maintain an internal mapping from one to the other in the `zcashd` build scripts as we deal in both worlds, but we identify with the LLVM / GCC ones as that's our outermost compilation surface. `zebra` is primarily Rust, so it would be fine to focus on the Rust-specific targets in its case.

Comment (gustavovalverde) [reply book/src/user/supported-platforms.md]: That's correct. Although it's common to do `gcc -dumpmachine` to know your actual `target triplet` (which matches with LLVM/GCC). I'm not aware if Rust is _adapting_ those, as I've been used to the common structure, which is stated here https://wiki.osdev.org/Target_Triplet 

We can surely indicate we're focusing on Rust-specific targets, as I don't think we're going a different way if we can comply with Rust's standards.

File: book/src/user/supported-platforms.md
Comment (gustavovalverde) [reply book/src/user/supported-platforms.md]: Fixed by [fix(docs): make rust version a separate column](https://github.com/ZcashFoundation/zebra/pull/4773/commits/f67eedccd648e74fbb4c286c3c189cc8f04b0e4b)

**Review by teor2345 — COMMENTED** [review]
I finished my detailed review. It looks really good, I think it's pretty clear about what we want.

I noticed some Rust compiler-specific language, and some unclear or incorrect details.

My suggestions to fix them are in PR #4791.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md:102]: Should tier 2 Zebra platforms require a specific Rust compiler tier?
Do we want to require Rust tier 1 or 2?

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md:164]: To meet these guarantees, tier 1 Zebra platforms require the platform to be a Rust compiler tier 1 platform.
Can we add that requirement?

File: book/src/user/target-tier-policies.md
Comment (gustavovalverde) [reply book/src/user/target-tier-policies.md:102]: This could (in the future) make our Tiers unpredictable, not making this document a single source of truth. But, at the same time, I kind of agree that this should be expressed somehow.

Isn't the `supported-platforms.md` clarifying this with the Rust versions column on each Tier?

File: book/src/user/supported-platforms.md
Comment (teor2345) [inline book/src/user/supported-platforms.md:36]: Sorry, I accidentally deleted beta Rust in my suggestion PR.
I've put it on a separate line so it's clearer.

```suggestion
| `x86_64-unknown-linux-gnu` | [GitHub ubuntu-latest](https://github.com/actions/virtual-environments#available-environments) | 64-bit | [latest stable release](https://github.com/rust-lang/rust/releases) | N/A
| `x86_64-unknown-linux-gnu` | [GitHub ubuntu-latest](https://github.com/actions/virtual-environments#available-environments) | 64-bit | [latest beta release](https://github.com/rust-lang/rust/blob/beta/src/version) | N/A
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]:
```suggestion
- Test failures on tier 2 platforms will be handled on a case-by-case basis.
  Depending on the severity of the failure, the Zebra team may decide to:
  - disable the test on that platform,
  - require a fix before the next release, or
  - remove the platform from tier 2.
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md]:
```suggestion
- The platform must build and pass tests reliably in CI, for all components that
  Zebra's CI considers mandatory.
  - Test failures on tier 1 platforms will be handled on a case-by-case basis.
    Depending on the severity of the failure, the Zebra team may decide to:
    - disable the test on that platform,
    - require a fix before the next release,
    - require a fix before any other PRs merge, or
    - remove the platform from tier 1.
  - The platform must not disable an excessive number of tests or pieces of tests
    in the testsuite in order to do so. This is a subjective requirement.
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [reply book/src/user/target-tier-policies.md]: I just made some suggestions to match our current process.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [reply book/src/user/target-tier-policies.md:102]: The Rust versions are different from the Rust platform tiers.

The same Rust stable release version is guaranteed to work on tier 1 platforms, and build on tier 2 platforms. So I guess we have to require tier 1, because a working Rust compiler is required to build Zebra.

File: book/src/user/target-tier-policies.md
Comment (teor2345) [inline book/src/user/target-tier-policies.md:131]:
```suggestion
- The platform must build reliably in CI, for all components that Zebra's CI
  considers mandatory.
  - Since a working Rust compiler is required to build Zebra,
    the platform must be a [Rust tier 1 platform](https://rust-lang.github.io/rustup-components-history/).
```

File: book/src/user/target-tier-policies.md
Comment (teor2345) [reply book/src/user/target-tier-policies.md:102]: I added some suggested working in tier 2 to cover this.

**Review by teor2345 — COMMENTED** [review]
Thank you for all your work tracking down these details.

I made a bunch of final suggestions to resolve all outstanding comments.
Feel free to accept or reject them, I think we're getting to the final details now.

**Review by teor2345 — APPROVED** [review]
Looks great, thanks!

Did you want to open a ticket to update the README and book indexes, or add it to the list of tasks in the release process ticket?

Things have changed a lot since our original release process draft. So the ticket and draft might be too complicated - we might want to delete the outdated parts before we start.

**Comment by gustavovalverde** [comment]
> Did you want to open a ticket to update the README and book indexes, or add it to the list of tasks in the release process ticket?

Added it to the list of tasks in the release process ticket.

**Pull request merged**
